pkg/controller/jobframework/setup.go

```diff
@@ -103,6 +110,64 @@ func (m *integrationManager) setupControllers(mgr ctrl.Manager, log logr.Logger,
 	})
 }
 
+func (m *integrationManager) setupControllerAndWebhook(mgr ctrl.Manager, name string, fwkNamePrefix string, cb IntegrationCallbacks, options Options, opts ...Option) error {
+	if err := cb.NewReconciler(
+		mgr.GetClient(),
+		mgr.GetEventRecorderFor(fmt.Sprintf("%s-%s-controller", name, options.ManagerName)),
+		opts...,
+	).SetupWithManager(mgr); err != nil {
+		return fmt.Errorf("%s: %w", fwkNamePrefix, err)
+	}
+	if err := cb.SetupWebhook(mgr, opts...); err != nil {
+		return fmt.Errorf("%s: unable to create webhook: %w", fwkNamePrefix, err)
+	}
+	m.enableIntegration(name)
+	return nil
+}
+
+func waitForAPI(ctx context.Context, mgr ctrl.Manager, log logr.Logger, gvk schema.GroupVersionKind, action func()) {
+	rateLimiter := workqueue.NewItemExponentialFailureRateLimiter(baseBackoffWaitForIntegration, maxBackoffWaitForIntegration)
+	item := gvk.String()
+	var defaultCheckAPIAvailable = func(mgr ctrl.Manager, gvk schema.GroupVersionKind) (bool, error) {
+		err := restMappingExists(mgr, gvk)
+		if meta.IsNoMatchError(err) {
+			return false, nil
+		}
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+
+	for {
+		isAvailable, err := defaultCheckAPIAvailable(mgr, gvk)
+		if isAvailable {
+			rateLimiter.Forget(item)
+			action()
+			return
+		}
+		if err != nil && !meta.IsNoMatchError(err) {
+			log.Error(err, "Failed to get REST mapping for gvk", "gvk", gvk)
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(rateLimiter.When(item)):
+			continue
+		}
+	}
+}
+
+func restMappingExists(mgr ctrl.Manager, gvk schema.GroupVersionKind) error {
+	mu.RLock()
```

Comment:
It looks like this is just for tests, so create the lock somewhere in the `MapperProvider` function, probably wrapping the Wrapper into another struct that has the lock.

Reply:
I've attempted to create a wrapper into another struct with the mapper and the lock, but the result is very similar.

The issue is this: In the application code, the go routine will continuously read the RESTMapping while in the test it will add to the mapper to simulate the availability of the newly added framework. It's a race condition, but of eventual consistency as the go routine reads again from the mapper and verifies the framework's availability.

I tried to lock the mapper in the test on its own, but the application code will continue to read from the mapper, causing this race condition.

Reply:
But in the application code, the mapper is updated without taking this lock (because it's defined somewhere else).

So the race condition will still exist.

Reply:
I think I understand your point, correct me please. If we were to add a new `CRD`, this would update the RESTMapper, we then at the same time will be requesting it to check/read for specific Mappings. If the RESTMapper has not yet been updated with the new CRD at the time we read from it, the mechanism is to re-read of which, is a race condition but of eventual synchronization. How do you think we could proceed?

Reply:
I don't know if the application RESTMapper itself has some internal lock already, it might be worth checking.

My point is that adding a lock here doesn't prevent a race condition in the application code, but maybe that's fine, eventually it will succeed.

Reply:
> My point is that adding a lock here doesn't prevent a race condition in the application code, but maybe that's fine, eventually it will succeed.

That's correct. You're right, the lock here is for the test to pass, given it's ran with the `-race` flag. Ideally, it's best to not introduce code here for the tests, but the `setup_test.go` (test file) essentially runs the `setup_controllers()` function which executes almost the entirety of the `setup.go` file, meaning it basically runs as a hybrid test or closely as it would run in production, hence requiring this lock here.